### ISSUE TITLE
Traces contains wrong HTTP status when unhandled exception is thrown

### ISSUE THREAD

**Issue by joshiste** [spring-projects/spring-boot#5331]
When an exception is thrown from a controller-method the Trace contains the HTTP status `200` instead of `500`. 
Spring Boot 1.3.3

I assume a similar fix for exceptions as in the `MetricsFilter` is necessary.
Additional to `status`, it would be nice to include `exception` and `message` in the traces too.

Output from `/trace`
```json
[{
  "timestamp": 1457044471670,
  "info": {
    "method": "GET",
    "path": "\/500",
    "headers": {
      "request": {
        "host": "localhost:8080",
        "user-agent": "curl\/7.43.0",
        "accept": "*\/*"
      },
      "response": {
        "X-Application-Context": "spring-boot-admin-sample",
        "status": "200"
      }
    }
  }
}]
```

 Controller:
```java
@Configuration
@EnableAutoConfiguration
@Controller
public class Application {
	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}
	@RequestMapping("/500")
	public void ex() {
		throw new IllegalStateException("foo");
	}
}
```

curl output:
```shell
$ curl -v http://localhost:8080/500
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /500 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Server: Apache-Coyote/1.1
< X-Application-Context: spring-boot-admin-sample
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Thu, 03 Mar 2016 22:34:31 GMT
< Connection: close
< 
* Closing connection 0
{"timestamp":1457044471690,"status":500,"error":"Internal Server Error","exception":"java.lang.IllegalStateException","message":"foo","path":"/500"}
```

**Comment by wilkinsona** [comment]
You haven't handled the exception in your application which means that as the request unwinds the status of the response is still `200`. It's only once the exception gets back into Tomcat that the status becomes `500`.

Handling the exception in your application and setting the response status will result in the trace reporting the status as 500. Like this, for example:

```
    @ExceptionHandler(IllegalStateException.class)
    public void handleException(HttpServletResponse response) {
    	response.setStatus(500);
    }
```

An alternative would be for Boot to auto-configure something that's container-specific (a Valve in the case of Tomcat) to record the traces. That wouldn't work in a war deployment though so I don't think it's a realistic option.

Thinking aloud, another option would be for the trace to be stored in a request attribute and for the error controller to update it once the container has redirected the request. That doesn't feel like a particularly good idea to me though.

In short, your best bet at the moment is to not rely on the container's exception handling and to ensure that the response status has been set before the request handling exits Spring MVC.

**Comment by joshiste** [comment]
The same applies to the `MetricFilter`. But it has code for the exception handling...

I wonder if there shouldn't be an default exception handling autoconfigured, so that we don't depend on container specific behaviour at all.

**Comment by wilkinsona** [comment]
> The same applies to the MetricFilter. But it has code for the exception handling...

Good point. We should do the same here.
